vars/template.go

```diff
@@ -130,50 +129,81 @@ func (i interpolator) extractVarNames(value string) []string {
 	var names []string
 
 	for _, match := range interpolationRegex.FindAllSubmatch([]byte(value), -1) {
-		names = append(names, strings.TrimPrefix(string(match[1]), "!"))
+		names = append(names, string(match[1]))
 	}
 
 	return names
 }
 
-type varsLookup struct {
-	varsTracker
-}
+func parseVarName(name string) VariableReference {
+	var pathPieces []string
 
-var ErrEmptyVar = errors.New("empty var")
+	varRef := VariableReference{Name: name}
 
-// Get value of a var. Name can be the following formats: 1) 'foo', where foo
-// is var name; 2) 'foo:bar', where foo is var source name, and bar is var name;
-// 3) '.:foo', where . means a local var, foo is var name.
-func (l varsLookup) Get(name string) (interface{}, bool, error) {
-	var splitName []string
 	if strings.Index(name, ":") > 0 {
-		parts := strings.Split(name, ":")
-		splitName = strings.Split(parts[1], ".")
-		splitName[0] = fmt.Sprintf("%s:%s", parts[0], splitName[0])
+		parts := strings.SplitN(name, ":", 2)
+		varRef.Source = parts[0]
+
+		pathPieces = pathRegex.FindAllString(parts[1], -1)
+
 	} else {
-		splitName = strings.Split(name, ".")
+		pathPieces = pathRegex.FindAllString(name, -1)
 	}
 
-	// this should be impossible since interpolationRegex only matches non-empty
-	// vars, but better to error than to panic
-	if len(splitName) == 0 {
-		return nil, false, ErrEmptyVar
+	varRef.Path = strings.ReplaceAll(pathPieces[0], "\"", "")
+	if len(pathPieces) >= 2 {
+		varRef.Fields = pathPieces[1:]
 	}
 
-	val, found, err := l.varsTracker.Get(splitName[0])
+	return varRef
+}
+
+// var ErrEmptyVar = errors.New("empty var")
+
+type varsTracker struct {
+	vars Variables
+
+	expectAllFound bool
+	expectAllUsed  bool
+
+	missing    map[string]struct{}
+	visited    map[string]struct{}
+	visitedAll map[string]struct{} // track all var names that were accessed
+}
+
+func newVarsTracker(vars Variables, expectAllFound, expectAllUsed bool) varsTracker {
+	return varsTracker{
+		vars:           vars,
+		expectAllFound: expectAllFound,
+		expectAllUsed:  expectAllUsed,
+		missing:        map[string]struct{}{},
+		visited:        map[string]struct{}{},
+		visitedAll:     map[string]struct{}{},
+	}
+}
+
+// Get value of a var. Name can be the following formats: 1) 'foo', where foo
+// is var name; 2) 'foo:bar', where foo is var source name, and bar is var name;
+// 3) '.:foo', where . means a local var, foo is var name.
```

Comment:
Comment needs updating, the main purpose now is to extract Fields from a `map[string]interface{}`

Reply:
I think the end goal of this `Get` is still finding a string value? Extracting fields is the first part. It then use the fields to do recursive look up.